### PR TITLE
RefreshWebcalService: randomly generate calendar-object uri

### LINKED ISSUE #19980: External ICS calendars fail to display if the UID includes slash
### Steps to reproduce
Create a file somewhere public with an http link to it with the following.

BEGIN:VCALENDAR
VERSION:2.0
PRODID: f2444c69b39463a8622ec7eab6bbe275                                                                                                   
BEGIN:VEVENT
UID:something/valid
DTSTAMP:20200310T000000
DTSTART:20200316T160000
DTEND:20200316T170000                                                                                                  
SUMMARY:Does not work
END:VEVENT                                                                                                                                   
END:VCALENDAR

Import that URL using the "new subscription from link" UI.

Try to view the event

### Expected behaviour
Event is shown, since the ICS is valid according to spec, as validated using https://icalendar.org/validator.html

### Actual behaviour
calendar indicator shows whirling spinner; never loads.

## Server configuration detail

**Operating system:** Linux 4.9.0-11-amd64 nextcloud/calendar#1 SMP Debian 4.9.189-3+deb9u1 (2019-09-20) x86_64

**Webserver:** nginx/1.10.3 (fpm-fcgi)

**Database:** mysql 10.1.41

**PHP version:** 

7.2.24-1+0~20191026.31+debian9~1.gbpbbacde
Modules loaded: Core, date, libxml, openssl, pcre, zlib, filter, hash, Reflection, SPL, sodium, session, standard, cgi-fcgi, mysqlnd, PDO, xml, calendar, ctype, curl, dom, mbstring, fileinfo, ftp, gd, gettext, iconv, igbinary, imagick, intl, json, exif, mysqli, pdo_mysql, Phar, posix, readline, redis, shmop, SimpleXML, sockets, sysvmsg, sysvsem, sysvshm, tokenizer, wddx, xmlreader, xmlwriter, xsl, zip, Zend OPcache, xdebug

**Nextcloud version:** 18.0.2 - 18.0.2.2

**Updated from an older Nextcloud/ownCloud or fresh install:** 

**Where did you install Nextcloud from:** unknown

<details><summary>Signing status</summary>

Array
(
)

</details>

<details><summary>List of activated apps</summary>

```
Enabled:
 - accessibility: 1.4.0
 - activity: 2.11.0
 - analytics: 2.1.1
 - apporder: 0.9.0
 - bookmarks: 2.3.4
 - bruteforcesettings: 1.5.0
 - calendar: 2.0.2
 - cloud_federation_api: 1.1.0
 - comments: 1.8.0
 - contacts: 3.2.0
 - dav: 1.14.0
 - deck: 0.8.0
 - federatedfilesharing: 1.8.0
 - files: 1.13.1
 - files_external: 1.9.0
 - files_fulltextsearch: 1.4.1
 - files_linkeditor: 1.0.13
 - files_pdfviewer: 1.7.0
 - files_rightclick: 0.15.2
 - files_sharing: 1.10.1
 - files_trashbin: 1.8.0
 - files_versions: 1.11.0
 - files_videoplayer: 1.7.0
 - fulltextsearch: 1.4.1
 - fulltextsearch_elasticsearch: 1.5.0
 - impersonate: 1.5.0
 - issuetemplate: 0.6.0
 - logreader: 2.3.0
 - lookup_server_connector: 1.6.0
 - news: 14.1.3
 - nextcloud_announcements: 1.7.0
 - notes: 3.2.0
 - notifications: 2.6.0
 - oauth2: 1.6.0
 - password_policy: 1.8.0
 - phonetrack: 0.6.2
 - photos: 1.0.0
 - polls: 1.3.0
 - previewgenerator: 2.2.0
 - privacy: 1.2.0
 - provisioning_api: 1.8.0
 - recommendations: 0.6.0
 - serverinfo: 1.8.0
 - settings: 1.0.0
 - spreed: 8.0.5
 - survey_client: 1.6.0
 - systemtags: 1.8.0
 - text: 2.0.0
 - theming: 1.9.0
 - theming_customcss: 1.5.0
 - twofactor_backupcodes: 1.7.0
 - updatenotification: 1.8.0
 - viewer: 1.2.0
 - workflowengine: 2.0.0
Disabled:
 - admin_audit
 - bookmarks_fulltextsearch
 - encryption
 - federation
 - firstrunwizard
 - sharebymail
 - support
 - user_ldap

```
</details>

<details><summary>Configuration (config/config.php)</summary>

```
{
    "instanceid": "***REMOVED SENSITIVE VALUE***",
    "passwordsalt": "***REMOVED SENSITIVE VALUE***",
    "secret": "***REMOVED SENSITIVE VALUE***",
    "trusted_domains": [
        "safe.artfulrobot.uk"
    ],
    "datadirectory": "***REMOVED SENSITIVE VALUE***",
    "overwrite.cli.url": "https:\/\/safe.artfulrobot.uk",
    "dbtype": "mysql",
    "version": "18.0.2.2",
    "dbname": "***REMOVED SENSITIVE VALUE***",
    "dbhost": "***REMOVED SENSITIVE VALUE***",
    "dbport": "",
    "dbtableprefix": "oc_",
    "mysql.utf8mb4": true,
    "dbuser": "***REMOVED SENSITIVE VALUE***",
    "dbpassword": "***REMOVED SENSITIVE VALUE***",
    "installed": true,
    "htaccess.RewriteBase": "\/",
    "overwritewebroot": "\/",
    "theme": "",
    "loglevel": 2,
    "maintenance": false,
    "updater.release.channel": "stable",
    "mail_from_address": "***REMOVED SENSITIVE VALUE***",
    "mail_smtpmode": "smtp",
    "mail_domain": "***REMOVED SENSITIVE VALUE***",
    "simpleSignUpLink.shown": false,
    "mail_sendmailmode": "smtp",
    "memcache.locking": "\\OC\\Memcache\\Redis",
    "memcache.local": "\\OC\\Memcache\\Redis",
    "redis": {
        "host": "***REMOVED SENSITIVE VALUE***",
        "port": 0,
        "dbindex": 0,
        "password": "***REMOVED SENSITIVE VALUE***",
        "timeout": 1.5
    },
    "app_install_overwrite": [
        "fulltextsearch_elasticsearch"
    ],
    "updater.secret": "***REMOVED SENSITIVE VALUE***"
}
```
</details>

**Are you using external storage, if yes which one:** local/smb/sftp/...

**Are you using encryption:** 

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...

## Client configuration

**Browser:** Mozilla/5.0 (X11; Linux x86_64; rv:75.0) Gecko/20100101 Firefox/75.0

**Operating system:** 

## Logs

<details><summary>Web server error log</summary>

```
none
```
</details>

<details><summary>Nextcloud log</summary>

```
nothing relevant at all
```
</details>

<details><summary>Browser log</summary>

REPORT https://safe.artfulrobot.uk/remote.php/dav/calendars/yyy/xxxxxxxxxx/

XHRREPORThttps://xxxxxxxxxxxxx/remote.php/dav/calendars/yyyy/xxxxxxxxx/
[HTTP/2 404 Not Found 146ms]

	
Response payload	

1

<?xml version="1.0" encoding="utf-8"?>

2

<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">

3

  <s:exception>Sabre\DAV\Exception\NotFound</s:exception>

4

  <s:message>Calendar object not found</s:message>

5

</d:error>

6

​

</details>

### PR DESCRIPTION
**Pull request by georgehrke**
fixes #19980

### CODE REVIEW

**Review by tcitworld — APPROVED** [review]

**Comment by kesselb** [comment]
`createCalendarObject` uses calendar id, object uri, uid and calendar type to look for existing objects. Does that still work if we use a random uri?

File: apps/dav/lib/CalDAV/WebcalCaching/RefreshWebcalService.php [hunk 1]
```diff
@@ -413,4 +412,13 @@ private function cleanURL(string $url) {
 
 		return $cleanURL;
 	}
+
+	/**
+	 * Returns a random uri for a calendar-object
+	 *
+	 * @return string
+	 */
+	public function getRandomCalendarObjectUri():string {
```
Comment (kesselb) [inline apps/dav/lib/CalDAV/WebcalCaching/RefreshWebcalService.php]: Could be protected.

**Comment by georgehrke** [comment]
@kesselb Yes, the current implementation is very simple. When refetching a webcal feed, it just deletes all cached events and stores the new ones.
https://github.com/nextcloud/server/pull/19986/files#diff-61d9e7310db26875b0eabc3a089d5cd9L112

**Review by kesselb — APPROVED** [review]

**Pull request merged**